File: engines/detail-engine/src/main/scala/io/sugo/pio/engine/ur/detail/DetailEngineFactory.scala
package io.sugo.pio.engine.ur.detail

import io.sugo.pio.engine.common.data.QueryableModelData
import io.sugo.pio.engine.ur.detail.data.{DetailModelData, DetailPreparedData, DetailTrainingData}
import io.sugo.pio.engine.ur.detail.engine._
import io.sugo.pio.spark.engine._
import io.sugo.pio.spark.engine.data.input.{BatchEventHose, PropertyHose}

class DetailEngineFactory(propertyHose: PropertyHose, batchEventHose: BatchEventHose) extends EngineFactory[DetailTrainingData, DetailPreparedData, DetailModelData, QueryableModelData] {
  override def createDatasource(): DataSource[DetailTrainingData] = {
    val dataSource = new DetailDataSource(propertyHose, batchEventHose)
    dataSource
  }

  override def createPreparator(): Preparator[DetailTrainingData, DetailPreparedData] = {
    val preparator = new DetailPreparator();
    preparator
  }

  override def createAlgorithm(): Algorithm[DetailPreparedData, DetailModelData] = {
    val algorithm = new DetailAlgorithm()
    algorithm
  }

  override def createModel(): Model[DetailModelData, QueryableModelData] = {
    val model = new DetailModel
    model
  }
}

File: engines/detail-engine/src/main/java/io/sugo/pio/engine/detail/DetailEngineFactory.java
package io.sugo.pio.engine.detail;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import io.sugo.pio.engine.data.input.BatchEventHose;
import io.sugo.pio.engine.data.output.Repository;
import io.sugo.pio.engine.detail.data.DetailModelData;
import io.sugo.pio.engine.detail.data.DetailPreparedData;
import io.sugo.pio.engine.detail.data.DetailTrainingData;
import io.sugo.pio.engine.training.*;

/**
 */
public class DetailEngineFactory implements EngineFactory {
    private final BatchEventHose batchEventHose;
    private final Repository repository;
    private final DetailEngineParams engineParams;

    @JsonCreator
    public DetailEngineFactory(@JsonProperty("batchEventHose") BatchEventHose batchEventHose,
                               @JsonProperty("repository") Repository repository,
                               @JsonProperty("engineParams") DetailEngineParams engineParams) {
        this.batchEventHose = batchEventHose;
        this.repository = repository;
        this.engineParams = engineParams;
    }

    @JsonProperty
    public BatchEventHose getBatchEventHose() {
        return batchEventHose;
    }

    @JsonProperty
    public Repository getRepository() {
        return repository;
    }

    @JsonProperty
    public DetailEngineParams getEngineParams() {
        return engineParams;
    }

    @Override
    public Engine createEngine() {
        return new DetailEngine(batchEventHose, repository, engineParams);
    }
}

